### PR TITLE
Collection Indexable functions

### PR DESCRIPTION
**Pull request by jayahariv**
* Add collection index functions & unit tests to validate. 
* Add collection management functions & unit tests to validate.
* CBLCollection internal constructor updated. 
* Kept database property in Scope & Collection. 
* Code cleanup - header include ordered

### CODE REVIEW

File: Objective-C/CBLCollection.mm [hunk 1]
```diff
@@ -21,45 +21,139 @@
...
+#import "CBLStatus.h"
+#import "CBLScope+Internal.h"
+
+using namespace fleece;
 
 NSString* const kCBLDefaultCollectionName = @"_default";
 
-@implementation CBLCollection
+@implementation CBLCollection {
+    __weak CBLDatabase* _db;
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:36]: This doesn't need to fix in this PR. The question is that how do you handle the case the CBLDatabase* pointer is dangle from the case that the database is released but the user still keep the collection object? When that happen, anything that uses _db in CBLCollection class will cause the crash including [db safeBlock:]. This is something that needs to be handled.

File: Objective-C/CBLCollection.mm [hunk 2]
```diff
@@ -21,45 +21,139 @@
...
+        CBLStringBytes sName(scopeName);
+        C4CollectionSpec spec = { .name = cName, .scope = sName };
+        
+        __block C4Collection* c;
+        __block C4Error err = {};
+        [db safeBlock:^{
+            c = c4db_getCollection(db.c4db, spec);
+            if (!c) {
+                // collection doesn't exists, create one
+                c = c4db_createCollection(db.c4db, spec, &err);
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:65]: 1. This logic is not correct for the case that you want to get the collection but don't want to create.
2. For the case of create, calling c4db_createCollection() is enough.

I would suggest that changing the signature to something as below, and let the CBLDatabase gives the c4collection.

`(instancetype) initWithDB: (CBLDatabase*)db collection: (C4Collection*)c4col error: (NSError**)error`.

File: Objective-C/CBLCollection.mm [hunk 3]
```diff
@@ -21,45 +21,139 @@
...
                        error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    CBLAssertNotNil(name);
+    CBLAssertNotNil(config);
+    
+    CBLDatabase* db = _db;
+    __block BOOL success = NO;
+    __block C4Error c4err;
+    [db safeBlock: ^{
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:104]: You would need to take care two things:
1. When the db pointer is not valid anymore.
2. When database is closed. So check Open is still needed.

Comment (pasin) [reply Objective-C/CBLCollection.mm:104]: Probably checking if the collection is valid might be enough? I think there is a PR in LiteCore that tries to handle this but not at the current master.

File: Objective-C/CBLCollection.mm (hunk 1 above)
Comment (pasin) [reply Objective-C/CBLCollection.mm:36]: Noted that if CBLDatabase doesn't cache any CBLCollections, you could retain db in CBLCollection.

File: Objective-C/CBLScope.m [hunk 4]
```diff
@@ -19,25 +19,31 @@
 
 #import "CBLScope+Internal.h"
 #import "CBLCollection+Internal.h"
+#import "CBLDatabase.h"
 
 NSString* const kCBLDefaultScopeName = @"_default";
 
-@implementation CBLScope
+@implementation CBLScope {
+    __weak CBLDatabase* _db;
```
Comment (pasin) [inline Objective-C/CBLScope.m:27]: Same comment as CBLCollection.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLCollection.mm (hunk 2 above)
Comment (jayahariv) [reply Objective-C/CBLCollection.mm:65]: CBLDatabase is not the only class which uses the [CBLCollection alloc] code, but CBLScope as well. So I guess, passing the collectionName and ScopeName can make it common NSString -> Slice conversion and everything inside CBLCollection itself.

File: Objective-C/CBLDatabase.mm [hunk 5]
```diff
@@ -722,45 +722,92 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
+    return [NSArray arrayWithArray: collections];
 }
 
-- (nullable CBLCollection*) createCollectionWithName: (NSString*)name
-                                      scope: (nullable NSString*)scope
-                                      error: (NSError**)error {
+- (nullable CBLCollection*) createCollectionWithName: (NSString*)collectionName
+                                               scope: (nullable NSString*)scopeName
+                                               error: (NSError**)error {
+    scopeName = scopeName.length > 0 ? scopeName : kCBLDefaultScopeName;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:755]: Just check nil scopeName for the default scope. If using nil, LiteCore will return an error when creating the collection.

File: Objective-C/CBLDatabase.mm [hunk 6]
```diff
@@ -722,45 +722,92 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
-    // TODO: add implementation
+    C4Collection* c4collection;
+    CBL_LOCK(self) {
+        [self mustBeOpen];
+        
+        C4Error c4err = {};
+        c4collection = c4db_createCollection(_c4db, spec, &c4err);
+        if (!c4collection) {
+            // TODO: handle error
+            return nil;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:768]: Let handle the error to complete the implementation.

File: Objective-C/CBLCollection.mm [hunk 7]
```diff
@@ -17,49 +17,122 @@
...
+@synthesize count=_count, name=_name, scope=_scope, db=_db;
 
-- (instancetype) initWithName: (NSString*)name
-                        scope: (nullable CBLScope*)scope
-                        error: (NSError**)error {
-    CBLAssertNotNil(name);
-    
+- (instancetype) initWithDB: (CBLDatabase*)db
+               c4collection: (C4Collection*)c4collection
+                      error: (NSError**)error {
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:44]: The error is not needed here. I have the same comment for the CBLScope's init as it is only for passing to the CBLScope's init.

File: Objective-C/CBLDatabase.mm [hunk 8]
```diff
@@ -722,45 +722,92 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
                             scope: (nullable NSString*)scope
                             error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    NSString* scopeName = scope ?: kCBLDefaultScopeName;
+    CBLStringBytes cName(name);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
+    CBL_LOCK(self) {
+        C4Error c4err = {};
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:808]: Need to check if the collection is valid before doing anything as well.

File: Objective-C/CBLCollection.mm (hunk 3 above)
Comment (jayahariv) [reply Objective-C/CBLCollection.mm:79]: collectionIsValid check done!

File: Objective-C/CBLDatabase.mm (hunk 8 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm:808]: does this mean, we have to get the c4collection from collectionSpec and then check validity and then delete? which is not very efficient right? instead LC can do the check, whether it exists, if not throw error? 

Or can we guarantee, getting a collection means, it exists and is valid. No need to further check for isValid?

Comment (jayahariv) [reply Objective-C/CBLDatabase.mm:828]: I see `c4db_hasCollection` API

**Comment by jayahariv** [comment]
Changed the `[db safeBlock:...]` code to directly use the `CBL_LOCK(db) { ... }`. Like you said, it started becoming messy, too many variables just collect the value to-from block.

File: Objective-C/Tests/DatabaseTest.m [hunk 9]
```diff
@@ -2421,22 +2421,159 @@ - (void) testDeleteIndex {
...
+// TODO: remove when implementation and unit tests are done
+// This is generic test to make sure, Collection APIs are working fine for QE.
 - (void) testCollection {
     [self createDocs: 10];
     CBLDocument* doc1 = [self.db documentWithID: $sprintf(@"doc_%03d", 1)];
     
     NSError* error = nil;
-    CBLDatabase* db = [[CBLDatabase alloc] initWithName: @"dbName" error: &error];
-    CBLCollection* c = [db createCollectionWithName: @"collection1" scope: nil error: &error];
+    CBLCollection* c = [self.db createCollectionWithName: @"collection1" scope: nil error: &error];
```
Comment (pasin) [inline Objective-C/Tests/DatabaseTest.m:2566]: What is the purpose of createDocs and get doc1 above?

File: Objective-C/CBLCollection.mm [hunk 10]
```diff
@@ -17,49 +17,128 @@
...
 - (BOOL) createIndexWithName: (NSString*)name
                       config: (CBLIndexConfiguration *)config
                        error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    CBLAssertNotNil(name);
+    CBLAssertNotNil(config);
+    
+    CBLDatabase* db = _db;
+    CBL_LOCK(db) {
```
Comment (pasin) [inline Objective-C/CBLCollection.mm:79]: I think it is probably time to change the way we do the lock but let's finish CBLCollection and CBLScope first.

I think we could switch to use NSLock / NSRecursiveLock. Another option is to use LiteCore's AccessLock.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/Tests/DatabaseTest.m [hunk 11]
```diff
@@ -2421,22 +2421,159 @@ - (void) testDeleteIndex {
...
+    AssertEqual(collections.count, 1);
+    AssertEqualObjects(collections[0].name, @"collection2");
+}
+
+- (void) testEmptyCollection {
+    AssertNil([self.db collectionWithName: @"dummy" scope: nil error: nil]);
+    AssertNil([self.db collectionWithName: @"dummy" scope: kCBLDefaultScopeName error: nil]);
+    AssertNil([self.db collectionWithName: @"dummy" scope: @"scope1" error: nil]);
+    
+    // TODO: When error is handled, check error here
```
[the remark this answers is not in the document]
Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m:2503]: will update, since we are returning error.

File: Objective-C/Tests/DatabaseTest.m (hunk 9 above)
Comment (jayahariv) [reply Objective-C/Tests/DatabaseTest.m:2566]: Sorry, this doc1 is used in below `[collection deleteDocument..]` functions we can remove when we are done with that.

File: Objective-C/CBLCollection.mm (hunk 10 above)
Comment (jayahariv) [reply Objective-C/CBLCollection.mm:79]: https://issues.couchbase.com/browse/CBL-3308

File: Objective-C/CBLDatabase.mm [hunk 12]
```diff
@@ -722,45 +726,131 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
+        if (![self mustBeOpen: error])
+            return nil;
+        
+        C4Error c4err = {};
+        FLMutableArray list = c4db_collectionNames(_c4db, sName, &c4err);
+        if (c4err.code != 0) {
+            CBLWarn(Database, @"%@ Failed to get collection names: %@ (%d/%d)",
+                    self, scopeName, c4err.domain, c4err.code);
+            convertError(c4err, error);
+            return nil;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:756]: Just notice this. The `list` needs to be released.

File: Objective-C/CBLDatabase.mm [hunk 13]
```diff
@@ -722,45 +726,131 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
+            NSString* name = FLValue_GetNSObject(FLArray_Get(list, i), nullptr);
+            
+            NSError* err = nil;
+            CBLCollection* c = [self collectionWithName: name scope: scopeName error: &err];
+            if (!c) {
+                CBLWarn(Database, @"%@ Failed to get collection: %@.%@ error: %@",
+                        self, scopeName, name, err);
+                if (error)
+                    *error = err;
+                return nil;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:772]: Just notice this. The `list` needs to be released.

File: Objective-C/CBLDatabase.mm [hunk 14]
```diff
@@ -722,45 +726,131 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
...
+            return nil;
+        }
+        
+        if (!c)
+            return nil;
+    }
     
-    CBLScope* s = [[CBLScope alloc] initWithName: scope ?: kCBLDefaultScopeName error: nil];
-    return [[CBLCollection alloc] initWithName: name scope: s error: nil];
+    return [[CBLCollection alloc] initWithDB: self c4collection: c];
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:836]: Just notice this, should be returned inside the lock to be able to retain `c` before the other thread could cause `c` to be freed on the other thread (remember that `c` is cached in LiteCore and could be released anytime).

**Review by pasin — CHANGES_REQUESTED** [review]
Notice a few more things. After fixing that, feel free to merge.

**Pull request merged**